frame/examples/tasks-example/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::marker::PhantomData;
+
+use codec::{Decode, Encode};
+use frame_support::dispatch::DispatchResult;
+// Re-export pallet items so that they can be accessed from the crate namespace.
+pub use pallet::*;
+use sp_runtime::DispatchError;
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[derive(Clone, PartialEq, Eq, Encode, Decode, TypeInfo)]
+	pub enum Task<T: Config> {
+		Increment,
+		Decrement,
+		#[doc(hidden)]
+		#[codec(skip)]
+		__Ignore(PhantomData<T>, frame_support::Never),
+	}
+
+	impl<T: Config> core::fmt::Debug for Task<T> {
+		fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+			f.debug_struct("Task").field("value", self).finish()
```

Comment:
Feels like this can be autogenerated by a macro.

Reply:
yeah a fair portion of that file is going to be auto-generated, that file is just to scaffold how the traits are laid out. I have actually already started some of the macros in my local branch